imageJPG.ml

```diff
@@ -17,71 +17,308 @@
+           let len = get_bytes ich 2 |> int_of_str2_be in
+           let data = get_bytes ich (len - 2) in
+           ignore data
+        );
+        find_next_marker ich handle_marker
     in
-    let (n, sof) = find_SOF headers in
-    (*
-    Printf.fprintf stderr "SOF%i: " n;
-    show_string_hex sof;
-    *)
-    let height = ((int_of_char sof.[1]) lsl 8) lor (int_of_char sof.[2]) in
-    let width = ((int_of_char sof.[3]) lsl 8) lor (int_of_char sof.[4]) in
-    width, height
+    try
+      find_next_marker ich handle_marker
+    with End_of_file ->
+      raise (Corrupted_image "Reached end of file while looking for SOF marker")
 
   let parsefile fn =
-    raise (Not_yet_implemented "ImageJPG.openfile") (* TODO  *)
+    let read_chunks = ref [] in
+
+    let sof = ref None in
+    let sos = ref None in
+    let image = ref None in
+
+    let rec handle_marker ich c =
+      let curr_ctype = string_of_marker c in
+      if !debug then
+        Printf.printf "Found marker %s\n%!" curr_ctype; (
+          match curr_ctype with
+        (* Required markers *)
+        | "SOI" ->
+          only_once ich read_chunks curr_ctype;
+          only_before ich read_chunks "SOS" curr_ctype;
+          is_first_chunk ich read_chunks curr_ctype;
+        | "EOI" ->
+          only_once ich read_chunks curr_ctype;
+          only_after ich read_chunks "SOI" curr_ctype;
+          only_after ich read_chunks "SOS" curr_ctype;
+          is_not_first_chunk ich read_chunks curr_ctype;
+
+          raise Exit
+        | "SOF0" | "SOF1"  | "SOF2"  | "SOF3"  | "SOF5"  | "SOF6"  | "SOF7"
+        | "SOF9" | "SOF10" | "SOF11" | "SOF13" | "SOF14" | "SOF15" as s ->
+          only_once ich read_chunks curr_ctype;
+          only_after ich read_chunks "SOI" curr_ctype;
+          only_before ich read_chunks "SOS" curr_ctype;
+
+          (* Frames other than SOF0 have patent issues, so are (currently) unimplemented *)
+          if s = "SOF0" then
+            let sof0 = parse_sof ich in
+            sof := Some sof0; (
+                Printf.printf "Number of components: %d\n%!" (Array.length sof0.components);
+                (*
+                 * JPEG requires being able to decode 1-4 components, but JFIF only requires
+                 * 1 and 3 components.
+                 *)
+                match Array.length sof0.components with
+                | 1 (* Y, AKA Greyscale *) ->
+                  image := Some (create_grey ~max_val:(ones sof0.precision) sof0.width sof0.height)
+                | 3 (* YCbCr, which will be converted to RGB *) ->
+                  image := Some (create_rgb ~max_val:(ones sof0.precision) sof0.width sof0.height)
+                | _ (* Handled in parse_sof *) -> 
+                  assert false
+                )
+        | "SOS" ->
+          only_after ich read_chunks "SOI" curr_ctype;
+
+          sos := Some (parse_sos ich)
+        | marker ->
+          Printf.printf "Marker %s is TODO\n%!" marker;
```

Comment:
remember `if !debug` please :)
EDIT: and `%S` if it can containg non-printable characters

Reply:
>remember if !debug please :)

Woops, thanks!

>EDIT: and %S if it can containg non-printable characters

Again, `string_of_marker` is always printable.